examples/step-39/step-39.cc

```diff
@@ -94,67 +94,191 @@ namespace Step39
+                  MeshWorker::IntegrationInfo<dim> &info) const override;
+    void face(MeshWorker::DoFInfo<dim>         &dinfo1,
+              MeshWorker::DoFInfo<dim>         &dinfo2,
+              MeshWorker::IntegrationInfo<dim> &info1,
+              MeshWorker::IntegrationInfo<dim> &info2) const override;
   };
 
 
-  // On each cell, we integrate the Dirichlet form. We use the library of
-  // ready made integrals in LocalIntegrators to avoid writing these loops
-  // ourselves. Similarly, we implement Nitsche boundary conditions and the
-  // interior penalty fluxes between cells.
+  // On each cell, we integrate the Dirichlet form as well as the
+  // Nitsche boundary conditions and the interior penalty fluxes between
+  // cells.
   //
   // The boundary and flux terms need a penalty parameter, which should be
-  // adjusted to the cell size and the polynomial degree. A safe choice of
-  // this parameter for constant coefficients can be found in
-  // LocalIntegrators::Laplace::compute_penalty() and we use this below.
+  // adjusted to the cell size and the polynomial degree. We compute it
+  // in two steps: First, we compute on each cell
+  // $K_i$ the value $P_i = p_i(p_i+1)/h_i$, where
+  // $p_i$ is the polynomial degree on cell $K_i$ and $h_i$ is the length of
+  // $K_i$ orthogonal to the current face. Second, if exactly one of the two
+  // cells adjacent to the face has children, its penalty is multiplied
+  // by two (to account for the fact that the mesh size $h_i$ there is
+  // only half that previously computed); it is possible that both adjacent
+  // cells are refined, in which case we are integrating over a non-active
+  // face and no adjustment is necessary. Finally, we return the average
+  // of the two penalty values.
   template <int dim>
-  void MatrixIntegrator<dim>::cell(
-    MeshWorker::DoFInfo<dim>                  &dinfo,
-    typename MeshWorker::IntegrationInfo<dim> &info) const
+  double ip_penalty_factor(const MeshWorker::DoFInfo<dim> &dinfo1,
+                           const MeshWorker::DoFInfo<dim> &dinfo2,
+                           unsigned int                    deg1,
+                           unsigned int                    deg2)
   {
-    LocalIntegrators::Laplace::cell_matrix(dinfo.matrix(0, false).matrix,
-                                           info.fe_values());
+    const unsigned int normal1 =
+      GeometryInfo<dim>::unit_normal_direction[dinfo1.face_number];
+    const unsigned int normal2 =
+      GeometryInfo<dim>::unit_normal_direction[dinfo2.face_number];
+    const unsigned int deg1sq = (deg1 == 0) ? 1 : deg1 * (deg1 + 1);
+    const unsigned int deg2sq = (deg2 == 0) ? 1 : deg2 * (deg2 + 1);
+
+    double penalty1 = deg1sq / dinfo1.cell->extent_in_direction(normal1);
+    double penalty2 = deg2sq / dinfo2.cell->extent_in_direction(normal2);
+    if (dinfo1.cell->has_children() && !dinfo2.cell->has_children())
+      penalty1 *= 2;
```

Comment:
I don't understand why this is different from the old code that does an XOR between the two bools. Your version is a easier to understand, though.

Reply:
Yes, I find XORs to be exceptionally difficult to read.

The difference is that in the original code, we only always updated `penalty1`, even if the second cell was refined. In my code here, I update `penalty1` above and `penalty2` below. It doesn't matter if you're on a uniform mesh, but it matters in general.

Reply:
Oh, gotcha!